tests/devices/test_default_qutrit.py

```diff
@@ -0,0 +1,574 @@
+    ],
+    dtype=np.complex128,
+)
+
+
+def include_inverses_with_test_data(test_data):
+    return test_data + [(item[0] + ".inv", item[1], item[2]) for item in test_data]
+
+
+def test_analytic_deprecation():
+    """Tests if the kwarg `analytic` is used and displays error message."""
+    msg = "The analytic argument has been replaced by shots=None. "
+    msg += "Please use shots=None instead of analytic=True."
+
+    with pytest.raises(
+        DeviceError,
+        match=msg,
+    ):
+        qml.device("default.qutrit", wires=1, shots=1, analytic=True)
+
+
+def test_dtype_errors():
+    """Test that if an incorrect dtype is provided to the device then an error is raised."""
+    with pytest.raises(DeviceError, match="Real datatype must be a floating point type."):
+        qml.device("default.qutrit", wires=1, r_dtype=np.complex128)
+    with pytest.raises(
+        DeviceError, match="Complex datatype must be a complex floating point type."
+    ):
+        qml.device("default.qutrit", wires=1, c_dtype=np.float64)
+
+
+# TODO: Add tests for expval, var, sample and tensor observables after addition of observables
+# TODO: Add tests to check for dtype preservation after more ops and observables have been added
+# TODO: Add tests for operations that will have custom internal implementations for default.qutrit once added
+# TODO: Add tests for inverse decomposition once decomposible operations are added
+# TODO: Add tests for verifying correct behaviour of different observables on default qutrit device.
+# TODO: Add tests for devices with shots for testing correct behaviour of sample
+
+
+class TestApply:
+    """Tests that operations and inverses of certain operations are applied correctly or that the proper
+    errors are raised.
+    """
+
+    # TODO: Add tests for non-parametric ops after they're implemented
+
+    # TODO: Add more data as parametric ops get added
+    test_data_single_wire_with_parameters = [
+        (qml.QutritUnitary, [1, 0, 0], [1, 1, 0] / np.sqrt(2), U_thadamard_01),
+        (qml.QutritUnitary, [1, 0, 0], [0, 0, 1], U_x_02),
+        (qml.QutritUnitary, [1, 0, 0], [1, 0, 0], U_z_12),
+        (qml.QutritUnitary, [0, 1, 0], [0, 1, 0], U_x_02),
+        (qml.QutritUnitary, [0, 0, 1], [0, 0, -1], U_z_12),
+        (qml.QutritUnitary, [0, 1, 0], [0, 0, 1], U_shift),
+        (qml.QutritUnitary, [0, 1, 0], [0, OMEGA, 0], U_clock),
+    ]
+
+    # TODO: Add more data as parametric ops get added
+    test_data_single_wire_with_parameters_inverse = [
+        (qml.QutritUnitary, [1, 0, 0], [0, 0, 1], U_shift),
+        (qml.QutritUnitary, [0, 0, 1], [0, 1, 0], U_shift),
+        (qml.QutritUnitary, [0, OMEGA, 0], [0, 1, 0], U_clock),
+        (qml.QutritUnitary, [0, 0, OMEGA**2], [0, 0, 1], U_clock),
+    ]
+
+    @pytest.mark.parametrize(
+        "operation, input, expected_output, par", test_data_single_wire_with_parameters
+    )
+    def test_apply_operation_single_wire_with_parameters(
```

Comment:
Okay so I get here that the qutrit_device_1_wire` and `tol` parameters are passed with some pytest magic through the fixtures. I think it would help for future developers looking at this that it is explained somewhere.

I find it a bit confusing to have that defined in a different file, does it have to be like that?

Reply:
The file in question, `tests/conftest.py` already existed before I started development on qutrits. I just added `qutrit_device_x_wires` for `x=1, 2, 3` to it while following the patterns used in `tests/devices/test_default_qutrit.py`. It definitely doesn't have to be like this, but I do see the benefit of having all fixtures in a single file to be used globally.

Reply:
Maybe then just add a small comment like "these tests are using fixtures, see (some resource what fixtures are), which are defined in <file where fixgures are defined, tests/conftest.py?>"

Reply:
Placing fixtures into `conftest.py` is a usual practice with `pytest`. Should be okay to leave it up to the reader to figure out (we use such fixtures at other places of the suite too).